feat: Add reusable LLM prompt template for SOL question generation

thatharshal opened the pull request:
<!-- Description -->
Adds a reusable prompt template for LLM-based generation of Select-One-from-Lot (SOL) questions in compliance with the schema and business rules defined in the ViBe platform.

**The template enforces:**

- Valid JSON-only response format
- Unique IDs and proper metadata
- Support for parameterized questions 
- Guidance for hints, explanations, time limits, and scoring
- Style parameters (subject, tone, grade level, etc.)

This will help generate high-quality, API-ready question payloads via LLM.

**Closes #317**

thatharshal committed:
feat: Add SOL question LLM prompt template, include bussiness logic rules and parameter handling, ensure JSON schema compliance

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,63 @@
+You are an expert question generator for the ViBe educational platform. Generate a single Select-One-from-Lot (SOL) question in JSON format following these exact specifications:
```
The role description could be broadened here. Referring to ViBe as an educational platform may not be entirely clear for an LLM. A more generalized role description would ensure better clarity and adaptability.

thatharshal replied:
Thank you for the review! Working on all the requested changes right away.

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+BUSINESS RULES:
```
Great job on defining the business rules! However, providing some examples for the LLM to refer to would enhance clarity. Including both correct and incorrect examples would give a clearer understanding of the expected behavior and help avoid any potential misinterpretations.

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,63 @@
+- Grade Level: ${gradeLevel}
+- Tone: ${tone} (formal|casual)
+
+Generate a question that tests conceptual understanding rather than mere factual recall.
```
This could come under a style parameter.

thatharshal replied:
@adityabmv I am not able to understand this request, please elaborate. This part is already under the Style parameter.

adityabmv replied:
Questions could be of two types with parameters/ without parameters.

This could be stated in style section.

thatharshal committed:
Update solQuestionTemplate.txt

Included the missing difficulty field and renamed a JSON field  (parameterized) to match the Question model (isParameterized).

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,63 @@
+   - points: 1–10 based on difficulty.
+6. Hint: Provide a pedagogically useful hint that guides thinking without revealing the answer.
+
+STYLE PARAMETERS:
```
It would be beneficial to provide a detailed explanation of each parameter along with examples of questions generated based on different combinations of those parameters. This would help clarify how the parameters influence the generated content and ensure that the system works as intended under various scenarios.

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,63 @@
+  "parameterized": boolean,
+  "parameters": [
+    {
+      "name": string,
```
Add a new field named `type`, which will accept primitive data types supported by JSON, such as `number`, `string`, and `boolean`. However, the value `null` should not be allowed for this field. This will help ensure that only valid, supported data types are used.

thatharshal committed:
Addressed the requested changes, and by adding detailed role, style parameter explanations and examples to prompt template

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

thatharshal commented:
@adityabmv  Please check the updated file for the requested changes.

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,189 @@
+
+
+2. Parameters: If parameterized=true:
+   - Each parameter name must appear exactly once in questionText as <QParam>{paramName}</QParam>.
```
Not exactly once, use atleast.

adityabmv commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,189 @@
+    - Example Values: "Grade 6", "Grade 10", "College", "Undergraduate"
+    - "gradeLevel": "Grade 6"
+
+- Tone: ${tone} (formal|casual)
```
You can remove tone, tone is not required. The tone should always be professional and formal.

thatharshal replied:
@adityabmv  Should i totally remove the tone parameter or keep the value to professional?

thatharshal committed:
Addressed updated comments

thatharshal commented on backend/ai/prompts/solQuestionTemplate.txt:
```diff
@@ -0,0 +1,189 @@
+    - Allowed Values: "professional"
+
+- Questions could be of two types with parameters or without parameters.
+- Generate a question that tests conceptual understanding rather than mere factual recall.
```
@adityabmv please check all the updates in file, for 2nd iteration of request.

thatharshal replied:
@adityabmv please check the changes

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template

thatharshal committed:
Merge branch 'main' into feature/llm-prompt-template
